esmvaltool/preprocessor/_derive/ohc.py

```diff
@@ -0,0 +1,48 @@
+"""Derivation of variable `ohc`."""
+import iris
+from iris import Constraint
+
+from cf_units import Unit
+
+from ._derived_variable_base import DerivedVariableBase
+
+RHO_CP = iris.cube.Cube(4.09169e+6, units=Unit('kg m-3 J kg-1 K-1'))
+
+
+class DerivedVariable(DerivedVariableBase):
+    """Derivation of variable `ohc`."""
+
+    # Required variables
+    _required_variables = {
+        'vars': [{'short_name': 'thetao', 'field': 'TO3M'}],
+        'fx_files': ['volcello']
+    }
+
+    def calculate(self, cubes,):
+        """
+        Compute ocean heat content.
+        Use c_p*rho_0= 4.09169e+6 J m-3 K-1 
+        (Kuhlbrodt et al., 2015, Clim. Dyn.)
+
+        Arguments
+        ---------
+        cube: iris.cube.Cube
+           input cube.
+
+        Returns
+        ---------
+        iris.cube.Cube
+              Output OHC cube.
+        """
+        # 1. Load the thetao and volcello cubes
+        cube = cubes.extract_strict(Constraint(cube_func=lambda c: 
+            c.var_name == 'thetao'))
+        volume = cubes.extract_strict(Constraint(cube_func=lambda c: 
+            c.var_name == 'volcello'))
+        # 2. multiply with each other and with cprho0
+        # try working with cubes here to save memory
+        # this might not be possible because of non-broadcasting in Iris
+#        cube *= volume * RHO_CP
+        cube.data *= volume.data * RHO_CP.data
```

Comment:
Do not realize the data unnessarily, see https://scitools.org.uk/iris/docs/latest/userguide/real_and_lazy_data.html
```suggestion
        cube *= volume * RHO_CP
```

Reply:
Indeed this is a work-around. The multiplication of the cubes did not work because (error message): "This operation cannot be performed as there are differing coordinates (longitude, depth, cell index along first dimension, latitude, cell index along second dimension) remaining which cannot be ignored." Of course, in this instance, cube is 4D, volume is 3D (for some models), and RHO_CP is a scalar. But since the dimensions match, it should work.
So broadcasting cubes doesn't work here for some reason.

Reply:
For this to be a meaningful operation, it is not enough that the dimensions match. Also the coordinates need to agree. For example you wouldn't want to multiply a 10x10 patch of cells at the equator with a 10x10 patch of cell areas from the arctic, nor a cube ten pressure levels deep with a field ten z levels deep. So iris is correct to refuse the operation here and it helpfully tells us which aspects are problematic, namely the three spatial dimensions. I would try it like this
```
cube *= RHO_CP
cube *= volume
```
just to confirm that the multiplication by `RHO_CP` is not problematic and then look at the differences between volume and the original cube.

Also, could you tell us with which dataset this fails?

Reply:
@zklaus is absolutely correct: @tillku see my comment below (am writing it now)